### PR TITLE
feat(components): Add mini slot map to shared components

### PR DESCRIPTION
**Pull request by Kadee80**
## overview

This PR addresses #4815 by adding the SVG mini slot map to the component library. 

Splitting up the component library and PD implementation into 2 separate PRs  for sanity sake. Thanks @IanLondon for the SVG pairing!

## changelog

- feat(components): Add mini slot map to shared components

## review requests

- [ ] component library examples match designs
- [ ] code looks sane

### CODE REVIEW

File: components/src/slotmap/SlotMap.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+const slotWidth = 33
+const slotHeight = 23
+const iconSize = 20
+
+export function SlotMap(props: SlotMapProps) {
+  const { collisionSlots, occupiedSlots, isError } = props
+  return (
+    <svg viewBox={`0,0,${slotWidth * 3}, ${slotHeight * 4}`}>
+      {SLOT_MAP_SLOTS.flatMap((row, rowIndex) =>
```
Comment (shlokamin) [inline components/src/slotmap/SlotMap.js]: Can we put the numbers 3 and 4 into constants like `DECK_WIDTH` + `DECK_HEIGHT` or something like that just to indicate what they mean?

File: components/src/slotmap/SlotMap.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+              />
+              {isCollisionSlot && (
+                <Icon
+                  className={styles.collision_icon}
+                  name="information"
+                  width={iconSize}
+                  height={iconSize}
+                  x={slotWidth / 2 - iconSize / 2}
+                  y={slotHeight / 2 - iconSize / 2}
+                />
```
Comment (shlokamin) [inline components/src/slotmap/SlotMap.js]: Sry if this is a dumb q but why do we do this? `slotWidth / 2 - iconSize / 2`

File: components/src/slotmap/SlotMap.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+export type SlotMapProps = {
+  /** Slot(s) to highlight */
+  occupiedSlots: Array<string>,
+  /** Optional collision warning */
+  collisionSlots?: Array<string>,
+  /** Optional error styling */
+  isError?: boolean,
+}
+
+export const SLOT_MAP_SLOTS = [
```
Comment (shlokamin) [inline components/src/slotmap/SlotMap.js]: It doesn't look like `SLOT_MAP_SLOTS` is being used outside of this file — why is it being exported?

File: components/src/slotmap/SlotMap.js (hunk 2 above)
Comment (Kadee80) [reply components/src/slotmap/SlotMap.js]: To center the icon (0,0 is on the top left corner)

File: components/src/slotmap/SlotMap.js (hunk 3 above)
Comment (Kadee80) [reply components/src/slotmap/SlotMap.js]: It might be in the follow up PR when we implement it in PD

**Comment by shlokamin** [comment]
A great suggestion by @b-cooper — might make more sense to pull in the deck def in `shared-date` and use them to render the svgs rather than introducing more sources of truth. Thoughts?

**Comment by IanLondon** [comment]
@shlokamin This is a super iconic/cartoony representation of the deck as 11 rectangles. The only kind of "truth" in it is the arrangement of slots in columns vs rows. But the column/row grid of slots is not directly represented in the deck definition, there's only a flat array `orderedSlots` of slots with X/Y coordinates. So we could drop the idea of column/row grid and try using those X/Y coords per slot, but we'd still have a problem: the gutter btw the slots is not present in this illustration, you'd have to do some tricky biz to alter those X/Y coordinates so that the gutter is gone. The gutter isn't a concept in the deck def, so you'd have to derive it also from X/Y coords and then afterwards, remove the gutter.

Alternative that is also kinda extra: you could construct the 2D array of slot names in grids (`SLOT_MAP_SLOTS`) from the deck def by going thru `orderedSlots` and using `x0 > x1` / `y0 > y1` comparisons to figure out how those slots are arranged in a grid.

For some future deck def, both these approaches could fail to make a useful iconic illustration here.

**Comment by shlokamin** [comment]
@IanLondon I hear your point about the gutter — I think getting around that would be annoying. Was curious about how this would look reusing `RobotWorkSpace` and after some playing around:

```
const { getDeckDefinitions } = require('./getDeckDefinitions')
const deckDef = getDeckDefinitions()['ot2_standard']
const slotName = '5'
const divSlot = '3'
;<RobotWorkSpace deckDef={deckDef} deckLayerBlacklist={[
  'calibrationMarkings',
  'fixedBase',
  'doorStops',
  'metalFrame',
  'removalHandle',
  'removableDeckOutline',
  'screwHoles',
  'slotNumbers',
  'slotRidges',
  'fixedTrash'
]}>
  {({ deckSlotsById }) => (
    <>
        {Object.keys(deckSlotsById).map(id => {
			const slot = deckSlotsById[id]
            if(!deckDef.locations.fixtures.some(f => f.slot === id)){ 
              return (
                <rect
                  x={slot.position[0]}
                  y={slot.position[1]}
                  width={slot.boundingBox.xDimension}
                  height={slot.boundingBox.yDimension}
                  fill="#0075ff33"
                  />
              )
			} else {
            	return null
            }
		})}
    </>
  )}
</RobotWorkSpace>
```

Here's what the screenshot looks like:



Screen Shot 2020-01-30 at 2 08 23 PM

This def doesn't totally match the design and you're right that the mini map is a more cartoon-y version of the real deck, it just seems a bit odd to diverge when the changes aren't super big. Totally understand if it's a design decision, just posting this here to show what it might look like if we did want to reuse `RobotWorkSpace` and use the real deck def.

**Comment by IanLondon** [comment]
@shlokamin the gutter is the problem, yup

Regarding future decks, it's not possible to guarantee that would work. It would render something, but is that the information we want to highlight to the user? Maybe, maybe not. That's hard to say without knowing what a future deck would be like and how we'd want to treat it differently than the existing deck. Like maybe a future deck would have slots that overlap each other, and a new iconic representation would be required. Or maybe it would pretty much work the same.

I don't think it's a good idea to tie design's hands about this. We have the CAD information for the deck in a JSON, that doesn't mean we must use it in all UI that has to do with the deck. For example, design has observed in user tests that when users see labware that is as physically accurate as possible, some of them will make wrong assumptions about it because it looks real. Design made the point that there are times we want things to look clearly non-real / iconic, because looking too real can mislead users. IDK if this is particular component is a case of that or not.

To me it feels like 2 very distinct concerns: iconic representations of the robot space that highlight information to communicate specific things to the user, versus physical dimensions of things on the deck used in situations where we want to reflect CAD file geometric relationships. IMO it's a "failure of separation of concerns" kinda antipattern to try to link iconic cartoons to CAD data.

**Pull request merged**